### ISSUE TITLE
valgrind reports 'potential memory leaks' from gflags

### ISSUE THREAD

**Issue by schuhschuh** [gflags/gflags#51]
Original [issue 40](https://code.google.com/p/gflags/issues/detail?id=40) created by schuhschuh on 2010-12-13T21:23:29.000Z:

<b>What steps will reproduce the problem?</b>
1. compile a program using gflags
2. run it with 'valgrind --leak_check=full'
3. look at all the memory leak reports from gflags

<b>What is the expected output? What do you see instead?</b>

expected: clean memory leak report
instead:  saw lots of potential memory leaks

These are not really memory leaks, but the presence of these leaks makes it harder to see real memory leaks.  Note that 'potential' memory leaks are sometimes quite dangerous.  For example, when working with Apache HTTPD, a large amount of memory allocation is using a pooled allocator.  You can leak memory into pools, and valgrind will report that as a 'potential leak'.

The gflags leak reports are not really problems like Apache pools, but the presence of them can obscure real problems.

<b>What version of the product are you using? On what operating system?</b>

http://google-gflags.googlecode.com/svn/tags/gflags-1.3/src (head)
OS: Linux Ubtuntu 10.04.

<b>Please provide any additional information below.</b>

**Comment by schuhschuh** [comment]
Incorrectly tagged commits 48e1a44d and cd1f401d related to this issue due to mismatch of issue ID between previous Google Code (#40) and this GitHub project. Updated bugfix branch to incorporated latest changes of master in preparation of final merge of bugfix into master (cf 470b3e88).

**Comment by schuhschuh** [comment]
Will introduce the changes of @MaxDZ8 with an increase of major version number due to API changes.

**Comment by schuhschuh** [comment]
Renamed related bugfix branch from "bugfix-40-memory-leaks" to [bugfix/#51-memory-leaks](https://github.com/gflags/gflags/tree/bugfix/%2351-memory-leaks) to remove confusion about issue number.

**Comment by schuhschuh** [comment]
I wonder if we should not get rid of the placement-new hack entirely when switching to non-POD string objects for string flags. We anyway break code which uses the string flags in a destructor of a static object. Take also note to include a warning in the documentation regarding this change and that global flag variables are not meant to be accessed in constructors/destructors. One can always pass the value of a flag after parsing the command line as member of a static object to circumvent the need for reading the value during static de-initializaiton. Disallowing such use would also allow us to use STL containers as flag types (cf. #82).

**Comment by ppedriana** [comment]
Are we going to see a fix for this bug? We are trying to do heap analysis here and this bug is getting in the way. Thanks.

**Comment by ppedriana** [comment]
Here is a proposed fix memory leak fix, which works in my testing with one compiler. It adds struct gflags_string_free_##name

```
#define DEFINE_string(name, val, txt)                                       \
  namespace fLS {                                                           \
    using ::fLS::clstring;                                                  \
    static union { void* align; char s[sizeof(clstring)]; } s_##name[2];    \
    clstring* const FLAGS_no##name = ::fLS::                                \
                                   dont_pass0toDEFINE_string(s_##name[0].s, \
                                                             val);          \
    static GFLAGS_NAMESPACE::FlagRegisterer o_##name(                       \
        #name, "string", MAYBE_STRIPPED_HELP(txt), __FILE__,                \
        s_##name[0].s, new (s_##name[1].s) clstring(*FLAGS_no##name));      \
    extern GFLAGS_DLL_DEFINE_FLAG clstring& FLAGS_##name;                   \
    using fLS::FLAGS_##name;                                                \
    clstring& FLAGS_##name = *FLAGS_no##name;                               \
                                                                            \
    struct gflags_string_free_##name{                                       \
        ~gflags_string_free_##name() {                                      \
            clstring* pstr0 = reinterpret_cast<clstring*>(s_##name[0].s);   \
            pstr0->~clstring();                                             \
            clstring* pstr1 = reinterpret_cast<clstring*>(s_##name[1].s);   \
            pstr1->~clstring();                                             \
        }                                                                   \
    };                                                                      \
    gflags_string_free_##name s_gflags_string_free_##name;                  \
  }                                                                         \
```

**Comment by schuhschuh** [comment]
@ppedriana Thanks for the patch. You're right that this should fix the string flags memory leaks. Not sure why I only tried to do so in the FlagValue destructor after I figured out that an explicit call of the destructor is needed... Given that the placement new is performed by this macro code, it's reasonable and good to have the corresponding "placement delete" part within the same code snippet/compilation unit.
